src/imp/lift_local_defs.rs

```diff
@@ -0,0 +1,129 @@
+use std::collections::BTreeMap;
+
+use indexmap::IndexMap;
+
+use crate::fun::{self, Name, Pattern};
+
+use super::{Definition, Expr, Stmt};
+
+impl Definition {
+  pub fn lift_local_defs(&mut self, gen: &mut usize) -> Result<IndexMap<Name, fun::Definition>, String> {
+    let mut defs = IndexMap::new();
+    self.body.lift_local_defs(&self.name, &mut defs, gen)?;
+    Ok(defs)
+  }
+}
+
+impl Stmt {
+  pub fn lift_local_defs(
+    &mut self,
+    parent: &Name,
+    defs: &mut IndexMap<Name, fun::Definition>,
+    gen: &mut usize,
+  ) -> Result<(), String> {
+    match self {
+      Stmt::LocalDef { .. } => {
+        let Stmt::LocalDef { mut def, mut nxt } = std::mem::take(self) else { unreachable!() };
+        let children = def.lift_local_defs(gen)?;
+        nxt.lift_local_defs(parent, defs, gen)?;
```

Comment:
These two lines will generate a strange order for the extracted functions. Nested definitions are lifted first, then the ones in the same depth are lifted bottom up.

Also, the generated names are difficult to understand, since local functions with the same name will have similar final names.
It would be a bit easier to read the output if they were named like this:
```py
def Foo():
  def A():
    def X():
  def B():
    def Y():
```
compiling to
```
Foo
Foo__local_0_A
Foo__local_0_A__local_0_X
Foo__local_1_B
Foo__local_1_B__local_0_Y
```

What you need to do is mostly just change the order of operations and reset `gen` for each layer of nested defs.